Module 2 Assignment - Part 1: One Layer NN with One Output
---

**Instructions:**

This assignment has a two large parts with many sub-parts. Each part will enable you to practice with a neural networks and to practice for Exam 2.

This Assignment will TAKE TIME so use all of the time you have to work on it. Please avoid waiting until the last few days as this will not be enough time. You need weeks - not days :)

Be sure to show all work and (when requested) to code using Python (and no NN/TF/Keras packages). Do not worry, we will use packages in coming modules but first it is best to learn about what is going on inside the model. 

You may use examples and code that I have shared as a reference. Using my code will not be considered cheating although I strongly recommend that you write as much (if not all) of your own code so that you learn the concepts more robustly. 

This is not a team assignment. Please work and code alone. You may discuss concepts, but you cannot share code or work with others. Papers that look too similar with split the grade (first offense) and can suffer less pleasant outcomes for further offenses. Keep it simple and smart - do your own work. 

![NN_architecture](network.jpg)

# 1) Fill out this network by hand and create all the needed matrices, etc. Show all multiplications and shapes.

In other words, illustrate the $X$, $y$, $W1$, $B$, $Z_1$, $H_1$, $W2$, $Z_2$, $C$, $\hat{y}$, $\hat{y}-y$, and $L$. Assume that $H_1$ uses the Sigmoid Function. 

As you do this - ask yourself:

- How many columns does $X$ have?
- How many hidden layers are there in this architecture?
- How many hidden units are there in the hidden layer?
- How many outputs are there?

## JK's answer:

### Input Matrix $X$

\begin{equation} 
X = 

\begin{bmatrix}
x_{11} & x_{12} & x_{13} \\
\vdots  & \vdots  & \vdots  \\
x_{n1} & x_{n2} & x_{n3} \\
\end{bmatrix}
\nonumber
\end{equation}

Where $n$ is any integer.

In any case, the point being: this architecture implies $X$ has 3 columns (but gives no restriction/information about the number of rows of $X$.)


### Targets $y$ 

\begin{align*}
\mathbf{y} = 
\begin{bmatrix} 
y_1 \\
\vdots \\
y_n
\end{bmatrix}
\end{align*}

### Weight matrix $W1$

The weight matrix connecting inputs $X$ with the first hidden layer (or in this case, the only hidden layer).

I will use both notations $W1 = W^{(1)}$, to be clear and be able to distinct weights $w$ in $W1$ and $W2$

\begin{align*}
W1 = W^{(1)} = 
\begin{bmatrix}
w_{11} & w_{12} & w_{13} & w_{14} \\
w_{21} & w_{22} & w_{23} & w_{24} \\
w_{31} & w_{32} & w_{33} & w_{34}
\end{bmatrix}
\end{align*}

When creating the matrix to match the subscript notation so that the first number in the subscript of $w$ is associated with the connected $x$ value and the second number is associated with the hidden unit $h$ (i.e. $w_{12}$ is the weight connecting $x_1$ to $h_2$), then the $W^{(1)}$ matrix is an $m \times p$ matrix, where $m$ is the number of columns of $X$ and $p$ is the number of units in the first hidden layer (which is also the number of rows of $H_1$). 

### Bias $B$

The "bias". Column vector with length that matches the number of rows in $X$.

$$\mathbf{B} = [b_1,\cdots, b_n]^\top$$

### $Z_1$

\begin{align*} 
Z_1 & =  X\cdot W_1 + B

\\
& = 
\begin{bmatrix}
x_{11} & x_{12} & x_{13} \\
\vdots  & \vdots  & \vdots  \\
x_{n1} & x_{n2} & x_{n3} \\
\end{bmatrix}

\begin{bmatrix}
w_{11} & w_{12} & w_{13} & w_{14} \\
w_{21} & w_{22} & w_{23} & w_{24} \\
w_{31} & w_{32} & w_{33} & w_{34}
\end{bmatrix}

+
\begin{bmatrix}
b_1 \\
\vdots \\
b_n
\end{bmatrix}

\\

& = 
\begin{bmatrix}
z_{11} & z_{12} & z_{13} & z_{14} \\
z_{21} & z_{22} & z_{23} & z_{24} \\
\vdots & \vdots & \vdots & \vdots \\
z_{n1} & z_{n2} & z_{n3} & z_{n4}
\end{bmatrix}
\end{align*}

Where we have:

\begin{align*}
& z_{11} = x_{11}w_{11} + x_{12}w_{21} + x_{13}w_{31} + b_1 & & \\
& z_{12} = x_{11}w_{12} + x_{12}w_{22} + x_{13}w_{32} + b_1 & &\\
\end{align*}

$$\vdots$$

\begin{align*}
z_{n4} = x_{n1}w_{14} + x_{n2}w_{24} + x_{n3}w_{34} + b_n 
\end{align*}

In other words, I guess any $Z$ element can be summarized as,

$$z_{j,k} = \sum_i^3 x_{j,i}w_{i,k} + b_j$$

### $H_1$

We're assuming the first hidden layer uses the Simoid activation function. Thus $S(z)$ is the Sigmoid of $z$. 

(Very) Generically we have,
$$h = S(z) = \frac{1}{1+e^{-z}}$$

Thus the matrix is as follows:

\begin{align*}
H_1 & = 
\begin{bmatrix}
S(z_{11}) & S(z_{12}) & S(z_{13}) & S(z_{14}) \\
\vdots    &   \vdots  &   \vdots  &   \vdots  \\
S(z_{n1}) & S(z_{n2}) & S(z_{n3}) & S(z_{n4}) \\
\end{bmatrix}

\\

& = 
\begin{bmatrix}
h_{11} & h_{12} & h_{13} & h_{14} \\
\vdots & \vdots & \vdots & \vdots \\
h_{n1} & h_{n2} & h_{n3} & h_{n4} \\
\end{bmatrix}
\end{align*}

The number of columns in $H_1$ is the number of hidden units in the hidden layer.

### $W2$

The weight matrix connecting the hidden layer and the last node.

Again, I will use both notations $W1 = W^{(1)}$ and $W2 = W^{(2)}$, to be able to distinguish weights from each weight matrix.

\begin{align*}
W2 = W^{(2)} =
\begin{bmatrix}
w_1^{(2)} \\[6pt]
w_2^{(2)} \\[6pt]
w_3^{(2)} \\[6pt]
w_4^{(2)} \\[6pt]
\end{bmatrix}
\end{align*}

### $C$ 

I believe this is the "bias" associated to calculate $Z_2$. So much like $B$ it's going to be a column vector with length that matches the number of rows in $X$.

\begin{equation*}
C = 
\begin{bmatrix}
c_1 \\
\vdots \\
c_n
\end{bmatrix}
\end{equation*}


### $Z_2$

\begin{align*}
Z_2 &= H_1 W^{(2)} + C \\
&= 
\begin{bmatrix}
h_{11} & h_{12} & h_{13} & h_{14} \\
\vdots & \vdots & \vdots & \vdots \\
h_{n1} & h_{n2} & h_{n3} & h_{n4} \\
\end{bmatrix}
\begin{bmatrix}
w_1^{(2)} \\[6pt]
w_2^{(2)} \\[6pt]
w_3^{(2)} \\[6pt]
w_4^{(2)} \\[6pt]
\end{bmatrix}

+

\begin{bmatrix}
c_1 \\
\vdots \\
c_n
\end{bmatrix}
\\
&= 
\begin{bmatrix}
h_{11}w_1^{(2)} + h_{12}w_2^{(2)} + h_{13}w_3^{(2)} + h_{14}w_4^{(2)} + c_1\\
\vdots \\
h_{n1}w_1^{(2)} + h_{n2}w_2^{(2)} + h_{n3}w_3^{(2)} + h_{n4}w_4^{(2)} + c_n
\end{bmatrix}
\\
&= 
\begin{bmatrix}
z_1^{(2)} \\
\vdots \\
z_n^{(2)}
\end{bmatrix}
\end{align*}

### $\mathbf{\hat{y}}$

In terms of notation,

\begin{align*}
\mathbf{\hat{y}} = 
\begin{bmatrix}
\hat{y}_1 \\
\vdots \\
\hat{y}_n \\
\end{bmatrix}
\end{align*}

(Should have the same shape as $\mathbf{y}$)

I'm guessing I have to do sigmoid again to get probability for binary labels?

\begin{align*}
\Rightarrow
\mathbf{\hat{y}} = 
\begin{bmatrix}
\hat{y}_1 \\
\vdots \\
\hat{y}_n \\
\end{bmatrix}
=
\begin{bmatrix}
S(z_1^{(2)}) \\
\vdots \\
S(z_n^{(2)})
\end{bmatrix}
\end{align*}

### $\mathbf{\hat{y}} - \mathbf{y}$

\begin{align*}
\mathbf{\hat{y}} - \mathbf{y} = 
\begin{bmatrix}
\hat{y}_1 \\
\vdots \\
\hat{y}_n \\
\end{bmatrix}

-

\begin{bmatrix} 
y_1 \\
\vdots \\
y_n
\end{bmatrix}

=

\begin{bmatrix}
\hat{y}_1 - y_1 \\
\vdots \\
\hat{y}_n - y_n \\
\end{bmatrix}

\end{align*}

### Loss $L$

There's really no limit of what Loss function $L$ to use. (A NN architecture doesn't define this, the NN model builder does :)) It could be LCE or a simple one.

For the purpose of this assignment I'll use a simple MSE (that's supposedly pretty commonly used as well), since it uses $\mathbf{\hat{y}} - \mathbf{y}$.

\begin{align*}
Loss = L &= \frac{1}{2} (\mathbf{y} - \mathbf{\hat{y}})^2 \\
&= \frac{1}{2}
\begin{bmatrix}
\hat{y}_1 - y_1 \\
\vdots \\
\hat{y}_n - y_n
\end{bmatrix}^2
\\
&=\frac{1}{2}
\begin{bmatrix}
\hat{y}_1 - y_1 \\
\vdots \\
\hat{y}_n - y_n
\end{bmatrix}
\begin{bmatrix}
\hat{y}_1 - y_1, & \cdots, & \hat{y}_1 - y_1
\end{bmatrix}
\\
&= \frac{1}{2} \sum_i^n (\hat{y}_i - y_i)^2
\end{align*}


# 2) Use Python to code the **Feed-Forward** portion of this NN. 

Print out $X$, $y$, $W1$, $B$, $Z_1$, $H_1$, $W_2$, $Z_2$, $C$, $\hat{y}$, $\hat{y}-y$, and $L$.

**Assumptions and Notes**:

(a) Create your own small labeled dataset. Have the right number of columns and have a column called "LABEL" which is 0 or 1. 

(b) Read in the data and create $X$ and $y$

(c) YOU create the dataset so that everyone will have a unique dataset. Have between 5 and 7 rows (no more), assure that your labels "make sense" for the data, and have a balance of labels (so nearish 50% $0$ and 50% $1$). You want your NN to "work" so again make sure your labeled data has a predictable pattern. *INVENT* the data yourself - type it into a `.csv` file - save the file as **A2_Data_YourName.csv**. You will submit this dataset with your Assignment submission.

(d) Create $H_1$ using Sigmoid. You will need to **code** a function for the Sigmoid (and its derivative) into your code. Do this using `def`.

(e) Use $L = \frac{1}{n} \displaystyle \sum^n (\hat{y} - y)^2$ as the Loss Function when $n$ is the number of rows in your dataset. Do not "hard code". The TAs or I should be able to use your code to read ANY dataset that has 3 columns of data, one label column called "LABEL" (of 0 or 1), and any number of rows. 

***JK answer:***

In [3]:
#%% libraries
import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
import math
from sklearn.preprocessing import LabelEncoder,MinMaxScaler
from sklearn.metrics import confusion_matrix  
import seaborn as sns
import random


In [4]:
# %%
class OneLayer_OneOutput_FF():
    def __init__(self,
                 data_filename = "A2_Data_JasmineKobayashi.csv", 
                 scale_data = True,
                 verbose = True):
        # Read Data
        self.df = pd.read_csv(data_filename)
        
        # create X and y, assuming data has column "LABEL" as column with labels
        ## X matrix
        if scale_data:
            mm_scaler = MinMaxScaler()
            X = mm_scaler.fit_transform(self.df.drop(columns="LABEL",axis=1))
        else:
            X = np.array(self.df.drop(columns=['LABEL'],axis=1))
        self.X = X

        ## y
        self.y = np.array(self.df[['LABEL']])

        if verbose:
            print("Shape of X:", self.X.shape)
            print("X matrix: \n", self.X)
            print("Shape of y:", self.y.shape)
            print("y vector: \n", self.y)

    def MSE_loss(self,y_hat,y):
        # L = mean((y_hat - y)^2)
        return np.mean((y_hat - y)**2)
        
    def lin_eq(self,X,W,B):
        assert X.shape[1] == W.shape[0], "Linear eq: z = X @ W + B. Shape of X: {}; Shape of W: {}".format(X.shape,W.shape)
        assert B.shape[0] == X.shape[0], "Linear eq: z = X @ W + B. Length of B should match X rows; Length B = {}; Shape of X = {}".format(B.shape[0],X.shape)
        return X @ W + B
    
    def Sigmoid(self,z,
                derivative = False):
        if derivative:
            return self.Sigmoid(z) * (1 - self.Sigmoid(z)) # dS/dz = S(z)(1-S(z))
        # Sigmoid
        return 1/(1 + math.e**(-z)) # S(z) = 1/(1 + e^(-z))
    
    def FeedForward(self,
                    hidden_units= 4,
                    randomize_parameters = False,
                    activation = "sigmoid",
                    verbose = True):
        # Shape of W1
        w1_ncol = hidden_units        # Number of W1 cols should be number of hidden units in H1
        w1_nrow = self.X.shape[1]     # Number of W1 rows should be number of X columns (dim of X)
        # Shape of B (column vector)
        b_length = self.X.shape[0]    # length of X rows
        # Shape of W2:
        w2_ncol = 1                   # Number of W2 cols should be number of outputs (in this case, only one output)
        w2_nrow = hidden_units        # Number of W2 rows should be number of hidden units in H1
        # Shape of C (column vector)
        c_length = self.X.shape[0]    # length of X rows
        
        if randomize_parameters: 
            # TODO: Finish potential code of randomized parameters
            # # bounds of randomized weights
            # bound = 1/np.sqrt(w1_nrow)   # I found online that supposedly this is a standard range for randomized weights
            # rand_w = [np.random.uniform(-bound,bound) for i in range(w1_ncol)]
            # self.W1 = np.array([rand_w]*w1_nrow)

            # self.W2 = np.array([]*w2_nrow)
            # print('All parameters were randomized')
            pass
        else:   # These are the parameters defined by the question 2 in Assignment2
            self.W1 = np.array([[1]*w1_ncol]*w1_nrow) # otherwise, all weights of W1 = 1
            self.B = np.array([[0]]*b_length)         # otherwise, all bias (B) = 0
            self.W2 = np.array([[2]*w2_ncol]*w2_nrow) # otherwise, all weights of W2 = 2
            self.C = np.array([[0]]*c_length)
            print("Parameters built for simple calculation (not randomized)")
        
        if verbose:
            # print("# of X cols (should be # of W1 rows):", self.X.shape[1])
            # print("# of hidden units (should be # of W1 cols):", hidden_units)
            print("W1 has shape:", self.W1.shape)
            print("W1 matrix: \n", self.W1)
            print("Shape of B:", self.B.shape)
            print("B (bias vector for Z1): \n", self.B)
            print("Shape of W2:",self.W2.shape)
            print("W2 matrix: \n", self.W2)
            print("Shape of C :", self.C.shape)
            print("C (bias vector for Z2): \n",self.C)

        # Create H1
        # Z1 = X @ W1 + B
        self.Z1 = self.lin_eq(self.X,self.W1,self.B)
        if verbose:
            print("Shape of Z1:",self.Z1.shape)
            print("Z1 matrix: \n",self.Z1)
        self.H1 = self.Sigmoid(self.Z1)
        # TODO: Add flexibility with other activation functions
        if verbose: 
            print("Activation function:",activation)
            print("H1 matrix: \n", self.H1)

        # Z2 and y-hat
        # Z2 = H1 @ W2 + C
        self.Z2 = self.lin_eq(self.H1,self.W2,self.C)
        self.y_hat = self.Sigmoid(self.Z2)

        if verbose:
            print("Shape of Z2:", self.Z2.shape)
            print("Z2 matrix: \n", self.Z2)
            print("y_hat:\n",self.y_hat)


In [7]:
NN = OneLayer_OneOutput_FF()
NN.FeedForward()
print("MSE Loss:",NN.MSE_loss(y_hat=NN.y_hat,y=NN.y))

Shape of X: (6, 3)
X matrix: 
 [[0.86363636 0.         0.07894737]
 [0.09090909 0.4        0.81578947]
 [0.         0.8        1.        ]
 [1.         0.2        0.05263158]
 [0.90909091 1.         0.        ]
 [0.04545455 0.6        0.94736842]]
Shape of y: (6, 1)
y vector: 
 [[0]
 [1]
 [1]
 [0]
 [0]
 [1]]
Parameters built for simple calculation (not randomized)
W1 has shape: (3, 4)
W1 matrix: 
 [[1 1 1 1]
 [1 1 1 1]
 [1 1 1 1]]
Shape of B: (6, 1)
B (bias vector for Z1): 
 [[0]
 [0]
 [0]
 [0]
 [0]
 [0]]
Shape of W2: (4, 1)
W2 matrix: 
 [[2]
 [2]
 [2]
 [2]]
Shape of C : (6, 1)
C (bias vector for Z2): 
 [[0]
 [0]
 [0]
 [0]
 [0]
 [0]]
Shape of Z1: (6, 4)
Z1 matrix: 
 [[0.94258373 0.94258373 0.94258373 0.94258373]
 [1.30669856 1.30669856 1.30669856 1.30669856]
 [1.8        1.8        1.8        1.8       ]
 [1.25263158 1.25263158 1.25263158 1.25263158]
 [1.90909091 1.90909091 1.90909091 1.90909091]
 [1.59282297 1.59282297 1.59282297 1.59282297]]
Activation function: sigmoid
H1 matrix: 
 

# 3) Compare *by-hand* and code

Here, will compare your by-hand work in (1) with your code in (2). 

To do this, use your dataset values (just the first row) and your label (for the first row) to FF *by hand on paper* through the network. Show numeric and clear results for $X$, $y$, $W_1$, $B$, $Z_1$, $H_1$, $W_2$, $Z_2$, $C$, $\hat{y}$, $\hat{y}-y$, and $L$.

Then, using your code and the same $X$ (first row) and $y$ (first label), print out $X$, $y$, $W_1$, $B$, $Z_1$, $H_1$, $W_2$, $Z_2$, $C$, $\hat{y}$, $\hat{y}-y$, and $L$.

Both the "by hand on paper" and the "by code" results should be the same. 

If they are not, find and fix your bugs until they are the same.

**Notes**: $\mathbf{W_1}$, $\mathbf{W_2}$, $\mathbf{B}$ (which is $b_1$, $b_2$, $b_3$, $b_4$), and $C$ (which is $c_1$) would normally be randomly initialized. However, to make your code and "by hand" comparison easier, let all the values in $W_1$ be $1$, let all the values in $W_2$ be 2, let all the $\mathbf{B}$ values be $0$, and let $c$ be $-1$. You will code this in and will use this by hand so that your code and by-hand results will be the same. 

Print out all key matrices and results. 

Illustrate and compare your "by-hand" results with your code print outs to show that they are the same. 

## JK's answer

### X Matrix

\begin{align*} 
X &= 

\begin{bmatrix}
x_{11} & x_{12} & x_{13} \\
x_{21} & x_{22} & x_{23} \\
\vdots  & \vdots  & \vdots  \\
x_{n1} & x_{n2} & x_{n3} \\
\end{bmatrix}
\\
&= 
\begin{bmatrix}
22 & 10 & 24 \\
5  & 12 & 52 \\
3  & 15 & 59 \\
25 & 11 & 23 \\
23 & 14 & 21 \\
4  & 13 & 57 
\end{bmatrix}
\end{align*}

In [8]:
NN = OneLayer_OneOutput_FF(scale_data=False,verbose=False)
print("non-scaled X:\n",NN.X)

non-scaled X:
 [[22 10 24]
 [ 5 12 52]
 [ 3 14 59]
 [25 11 23]
 [23 15 21]
 [ 4 13 57]]


To avoid errors with Sigmoid, so going to Min-Max scale the data. I'm not going to calculate individually I"m going to leave it to the Sklearn package min-max normalizer and trust it works.

But this is the formula I would use if I did calculate by hand:

$$x' = \frac{x - min}{max - min}$$

Where $x'$ is the scaled value, $x$ is the value to scale. ($min$ & $max$ are the minimum value and maximum value of the list of values to scale to.)

In [9]:
NN = OneLayer_OneOutput_FF(scale_data=True,verbose=False)
print("Min-Max Scaled X:\n",NN.X)

Min-Max Scaled X:
 [[0.86363636 0.         0.07894737]
 [0.09090909 0.4        0.81578947]
 [0.         0.8        1.        ]
 [1.         0.2        0.05263158]
 [0.90909091 1.         0.        ]
 [0.04545455 0.6        0.94736842]]


### Targets $\mathbf{y}$

\begin{align*}
\mathbf{y} &= 
\begin{bmatrix} 
y_1 \\
\vdots \\
y_n
\end{bmatrix}
\\
&=
\begin{bmatrix} 
0 \\
1 \\
1 \\
0 \\
0 \\
1
\end{bmatrix}
\end{align*}

In [12]:
print("y target vector:\n",NN.y)

y target vector:
 [[0]
 [1]
 [1]
 [0]
 [0]
 [1]]


### $W^{(1)}$ Matrix
For this assignment we're using all weights in $W^{(1)}$ are equal to $1$

\begin{align*}
W^{(1)} &= 
\begin{bmatrix}
w_{11} & w_{12} & w_{13} & w_{14} \\
w_{21} & w_{22} & w_{23} & w_{24} \\
w_{31} & w_{32} & w_{33} & w_{34}
\end{bmatrix}
\\
&=
\begin{bmatrix}
1 & 1 & 1 & 1 \\
1 & 1 & 1 & 1 \\
1 & 1 & 1 & 1
\end{bmatrix}
\end{align*}

In [14]:
NN.FeedForward(verbose=False)
print("W1 matrix: \n",NN.W1)

Parameters built for simple calculation (not randomized)
W1 matrix: 
 [[1 1 1 1]
 [1 1 1 1]
 [1 1 1 1]]


### Bias vector $\mathbf{B}$

For this assignment we using all bias = 0

\begin{align*}
\mathbf{B} =
\begin{bmatrix}
b_1 \\
\vdots \\
b_n
\end{bmatrix}
=
\begin{bmatrix}
0 \\
0 \\
0 \\
0 \\
0 \\
0 \\
\end{bmatrix}
\end{align*}

In [15]:
print("Bias vector B: \n",NN.B)

Bias vector B: 
 [[0]
 [0]
 [0]
 [0]
 [0]
 [0]]


### $Z_1$ Matrix

\begin{align*}
Z_1 &= X \cdot W^{(1)} + B \\
&= 
\begin{bmatrix}
x_{11} & x_{12} & x_{13} \\
x_{21} & x_{22} & x_{23} \\
\vdots  & \vdots  & \vdots  \\
x_{n1} & x_{n2} & x_{n3} \\
\end{bmatrix}
\begin{bmatrix}
w_{11} & w_{12} & w_{13} & w_{14} \\
w_{21} & w_{22} & w_{23} & w_{24} \\
w_{31} & w_{32} & w_{33} & w_{34}
\end{bmatrix}
+
\begin{bmatrix}
b_1 \\
\vdots \\
b_n
\end{bmatrix}
\\
&= 
\begin{bmatrix}
0.863 & 0.0 & 0.0789 \\
0.091 & 0.4 & 0.816  \\
0.0   & 0.8 & 1.0    \\
1.0   & 0.2 & 0.0526 \\
0.909 & 1.0 & 0.0    \\
0.045 & 0.6 & 0.947 
\end{bmatrix}
\begin{bmatrix}
1 & 1 & 1 & 1 \\
1 & 1 & 1 & 1 \\
1 & 1 & 1 & 1
\end{bmatrix}
+
\begin{bmatrix}
0 \\
0 \\
0 \\
0 \\
0 \\
0 \\
\end{bmatrix}
\\
&=
\begin{bmatrix}
0.942 & 0.942 & 0.942 & 0.942 \\
1.306 & 1.306 & 1.306 & 1.306 \\
1.8   & 1.8   & 1.8   & 1.8   \\
1.253 & 1.253 & 1.253 & 1.253 \\
1.91  & 1.91  & 1.91  & 1.91  \\
1.597 & 1.597 & 1.597 & 1.597 
\end{bmatrix}
\end{align*}

In [16]:
NN.Z1

array([[0.94258373, 0.94258373, 0.94258373, 0.94258373],
       [1.30669856, 1.30669856, 1.30669856, 1.30669856],
       [1.8       , 1.8       , 1.8       , 1.8       ],
       [1.25263158, 1.25263158, 1.25263158, 1.25263158],
       [1.90909091, 1.90909091, 1.90909091, 1.90909091],
       [1.59282297, 1.59282297, 1.59282297, 1.59282297]])

### $H_1$ Matrix

Using the Sigmoid as the activation function which is defined as follows,

\begin{align*}
S(z) &= \frac{e^z}{1+e^z} \\[6pt] 
&= \frac{1}{1+e^{-z}}
\end{align*}

And $h=S(z)$

\begin{align*}
H_1 & = 
\begin{bmatrix}
S(z_{11}) & S(z_{12}) & S(z_{13}) & S(z_{14}) \\
\vdots    &   \vdots  &   \vdots  &   \vdots  \\
S(z_{n1}) & S(z_{n2}) & S(z_{n3}) & S(z_{n4}) \\
\end{bmatrix}

\\

& = 
\begin{bmatrix}
0.7195 & 0.7195 & 0.7195 & 0.7195 \\
0.787  & 0.787  & 0.787  & 0.787  \\
0.858  & 0.858  & 0.858  & 0.858  \\
0.778  & 0.778  & 0.778  & 0.778  \\
0.871  & 0.871  & 0.871  & 0.871  \\
0.832  & 0.832  & 0.832  & 0.832  
\end{bmatrix}
\end{align*}

In [17]:
print("H1 matrix: \n",NN.H1)

H1 matrix: 
 [[0.71962126 0.71962126 0.71962126 0.71962126]
 [0.78696018 0.78696018 0.78696018 0.78696018]
 [0.85814894 0.85814894 0.85814894 0.85814894]
 [0.77775507 0.77775507 0.77775507 0.77775507]
 [0.87091698 0.87091698 0.87091698 0.87091698]
 [0.8310129  0.8310129  0.8310129  0.8310129 ]]


### $W^{(2)}$ Matrix

For this assignment we're using all weights in $W^{(2)}$ are equal to $2$

\begin{align*}
W^{(2)} =
\begin{bmatrix}
w_1^{(2)} \\[6pt]
w_2^{(2)} \\[6pt]
w_3^{(2)} \\[6pt]
w_4^{(2)} \\[6pt]
\end{bmatrix}
=
\begin{bmatrix}
2 \\
2 \\
2 \\
2 
\end{bmatrix}
\end{align*}

In [18]:
print("W2 matrix: \n",NN.W2)

W2 matrix: 
 [[2]
 [2]
 [2]
 [2]]


### $\mathbf{C}$ bias vector

For this assignment we using all bias = 0

\begin{equation*}
C = 
\begin{bmatrix}
c_1 \\
\vdots \\
c_n
\end{bmatrix}
=
\begin{bmatrix}
0 \\
0 \\
0 \\
0 \\
0 \\
0 
\end{bmatrix}
\end{equation*}

In [19]:
print("C bias vector: \n",NN.C)

C bias vector: 
 [[0]
 [0]
 [0]
 [0]
 [0]
 [0]]


### $Z_2$ Matrix

\begin{align*}
Z_2 &= H_1 W^{(2)} + C \\
&= 
\begin{bmatrix}
h_{11} & h_{12} & h_{13} & h_{14} \\
\vdots & \vdots & \vdots & \vdots \\
h_{n1} & h_{n2} & h_{n3} & h_{n4} \\
\end{bmatrix}
\begin{bmatrix}
w_1^{(2)} \\[6pt]
w_2^{(2)} \\[6pt]
w_3^{(2)} \\[6pt]
w_4^{(2)} \\[6pt]
\end{bmatrix}

+

\begin{bmatrix}
c_1 \\
\vdots \\
c_n
\end{bmatrix}

= 
\begin{bmatrix}
z_1^{(2)} \\
\vdots \\
z_n^{(2)}
\end{bmatrix}
\\
&= 
\begin{bmatrix}
0.7195 & 0.7195 & 0.7195 & 0.7195 \\
0.787  & 0.787  & 0.787  & 0.787  \\
0.858  & 0.858  & 0.858  & 0.858  \\
0.778  & 0.778  & 0.778  & 0.778  \\
0.871  & 0.871  & 0.871  & 0.871  \\
0.832  & 0.832  & 0.832  & 0.832  
\end{bmatrix}
\begin{bmatrix}
2 \\
2 \\
2 \\
2 
\end{bmatrix}
+
\begin{bmatrix}
0 \\
0 \\
0 \\
0 \\
0 \\
0 
\end{bmatrix}
\\
&= 
\begin{bmatrix}
5.756 \\
6.296 \\
6.864 \\
6.224 \\
6.968 \\
6.656
\end{bmatrix}
\end{align*}

In [20]:
print("Z2 matrix: \n",NN.Z2)

Z2 matrix: 
 [[5.75697011]
 [6.29568146]
 [6.86519148]
 [6.22204054]
 [6.96733585]
 [6.64810323]]


#### $\mathbf{\hat{y}}$ prediction vector

I'm guessing I have to do sigmoid again to get probability for binary labels.

So something like,

\begin{align*}
\mathbf{\hat{y}} &= 
\begin{bmatrix}
\hat{y}_1 \\
\vdots \\
\hat{y}_n \\
\end{bmatrix}
=
\begin{bmatrix}
S(z_1^{(2)}) \\
\vdots \\
S(z_n^{(2)})
\end{bmatrix}
\\
& = 
\begin{bmatrix}
0.997 \\
0.998 \\
0.999 \\
0.998 \\
0.999 \\
0.999 
\end{bmatrix}
\end{align*}

In [21]:
print("y-hat prediction vector:\n",NN.y_hat)

y-hat prediction vector:
 [[0.99684928]
 [0.99815914]
 [0.99895761]
 [0.99801874]
 [0.99905873]
 [0.9987052 ]]


### MSE Loss

We're using MSE

\begin{align*}
L = \frac{1}{n} \displaystyle \sum_i^n (\hat{y}_i - y_i)^2
\end{align*}

Gonna start with $\hat{y} - y$

\begin{align*}
\mathbf{\hat{y}} - \mathbf{y} &= 
\begin{bmatrix}
\hat{y}_1 \\
\vdots \\
\hat{y}_n \\
\end{bmatrix}

-

\begin{bmatrix} 
y_1 \\
\vdots \\
y_n
\end{bmatrix}

=

\begin{bmatrix}
\hat{y}_1 - y_1 \\
\vdots \\
\hat{y}_n - y_n \\
\end{bmatrix}
\\
&=
\begin{bmatrix}
0.997 \\
0.998 \\
0.999 \\
0.998 \\
0.999 \\
0.999 
\end{bmatrix}
-
\begin{bmatrix} 
0 \\
1 \\
1 \\
0 \\
0 \\
1
\end{bmatrix}
=
\begin{bmatrix}
0.997  \\
-0.002 \\
-0.001 \\
0.998  \\
0.999  \\
-0.001 
\end{bmatrix}
\end{align*}

In [22]:
print("(y_hat) - y:\n", NN.y_hat-NN.y)

(y_hat) - y:
 [[ 0.99684928]
 [-0.00184086]
 [-0.00104239]
 [ 0.99801874]
 [ 0.99905873]
 [-0.0012948 ]]


\begin{align*}
\Rightarrow
L  \frac{1}{n} \displaystyle \sum_i^n (\hat{y}_i - y_i)^2
\approx 0.498
\end{align*}

In [24]:
print("MSE Loss:",NN.MSE_loss(y_hat=NN.y_hat,y=NN.y))

MSE Loss: 0.4979790666878168


# 4) Writing out the derivatives by hand.

For this part and in order to help you to code the back propagation, write out the derivatives for updating $\mathbf{W_1}$, $\mathbf{W_2}$, $\mathbf{B}$, and $C$.

Please write them clearly :)

You can do this "matrix style" - so for example, $\frac{\partial L}{\partial \mathbf{W_1}}$, or you can write them out individually, such as $\frac{\partial L}{dw_{11}}$, $\frac{\partial L}{\partial w_{12}}$, etc...

Either way, show all your work and illustrate how the matrices (such as $\mathbf{X}$, $\mathbf{W_1}$, $\mathbf{W_2}$, etc. etc.) will all fit together to create the final derivative.

**Hints:** There are many examples on the PowerPoints. We reviewed this in class. You can also see examples of this in my code examples which are all here: [Gates Bolton Analytics - Neural Networks Page](https://gatesboltonanalytics.com/?page_id=680)

## JK answer:



# 5) Adding Back Propagation to your code so that $\mathbf{W_1}$, $\mathbf{B}$, and $C$ are updated.

For this step, you will add back propagation to your code. Specifically, you already have code **(2)** above. In your code (so far) you have read in your dataset, have created $X$ and $y$, and have function (or class if you OO programming) for Feed Forward.

In this next step, you will add a function for Back Propagation that will update your parameters. Recall that the parameters for this NN architecture are $\mathbf{W_1}$, $\mathbf{W_2}$, $\mathbf{B}$, and $C$. Each must be updated (using a learning rate) and the derivatives you calculated in **(4)**.

Once you create code for the Back Propagation, add further code for epochs so that your code will "update" (FF and BP and calculate the Loss each time) for as many iterations (epochs) as you wish.

Be sure to have a visualization of the change in the average Loss over epochs. 

Be sure to have a final visualization that is a confusion matrix that shows the label predictions.

Be sure to **print out** some of your matrices so that a viewer can see what you are doing. While there are many ways to do this, you can consider printing $\mathbf{W_1}$, the updates for $\mathbf{W_1}$, the new $\mathbf{W_1}$ (which will be the derivative updates times the learning rate added to $\mathbf{W_1}$). In other words, **make your code verbose (shows/tells what its doing)** so you can print out the last epoch steps. (An example on Canvas) 


**NOTE:** At this point, you should have code that will read in any dataset with 3 columns of numeric data and one column called "LABEL" (of $0$ or $1$) and will perform NN modeling and a prediction result on the dataset.

***JK answer:***

# 6) Test model

Finally, create a test dataset and use your model to predict the labels of that test dataset. Then create a confusion matrix to show the prediction accuracy. Remember, to do this you will need labeled data, BUT you will remove the labels from the data before trying to predict it. Once the model makes predictions, you can then compare the predictions to the actual labels. 

(1) EXPLAIN in words what you plan to do

(2) Code it

(3) Give the results as a confusion matrix

(4) Include your test dataset with your submission.

***JK answer:***